### PR TITLE
[FLINK-35858][Runtime/State] Namespace support for async state

### PR DESCRIPTION
**Pull request by Zakelly**
## What is the purpose of the change

This PR add namespace for async state, for partitioning the state for internal usage.

## Brief change log

 - Add type parameter for internal state APIs.
 - code path for creating state with namespace.
 - Forst support

## Verifying this change

Some related existing tests have been modified.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): yes
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

**Review by yunfengzhou-hub — COMMENTED** [review]
Thanks for the PR! LGTM.

File: flink-runtime/src/main/java/org/apache/flink/runtime/asyncprocessing/RecordContext.java [hunk 1]
```diff
@@ -56,6 +59,9 @@ public class RecordContext<K> extends ReferenceCounted<RecordContext.DisposerRun
     /** The keyGroup to which key belongs. */
     private final int keyGroup;
 
+    /** The namespaces of states. Lazy initialization for saving memory. */
```
Comment (fredia) [inline flink-runtime/src/main/java/org/apache/flink/runtime/asyncprocessing/RecordContext.java:62]: Will there be any concurrency problem caused by ` Lazy initialization`?

**Review by fredia — COMMENTED** [review]
Thanks for the PR, overall LGTM except for one minor comment.

File: flink-runtime/src/main/java/org/apache/flink/runtime/asyncprocessing/RecordContext.java (hunk 1 above)
Comment (Zakelly) [reply flink-runtime/src/main/java/org/apache/flink/runtime/asyncprocessing/RecordContext.java:62]: Nope. The `setNamespace` is only invoked on task thread.

**Review by fredia — APPROVED** [review]

**Comment by Zakelly** [comment]
Thanks for the review, merging...

**Pull request merged**